### PR TITLE
[WIP] JAX implementation

### PR DESCRIPTION
**Pull request by bamos**
Hey! Hope you all have been doing well. I've been using JAX and was finally able to connect this in for #7. This initial version is pretty similar to the PyTorch one and was easy with some of the recent JAX developments on enabling custom JAX primitives with a non-traced vjp. The interface follows our PyTorch/TF layers has an outer function that takes the CVXPY objects and then returns a callable into a primitive for the problem.

Let me know if you have any thoughts on this initial version! There are a few things left I'll finish up over the next few days:

+ [x] Add DGP
+ [x] Add the rest of the PyTorch tests and add JAX to the CI
+ [x] Update the README (image and JAX example)
+ [x] Document current limitations somewhere (e.g. can't use this with the JIT/vmap, and can't do higher-order derivatives)
+ [x] Add a sample notebook

Some useful JAX references:
+ https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html
+ https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html
+ https://github.com/google/jax/issues/1142 and https://github.com/google/jax/issues/3415.  @mattjj and @shoyer, your responses in these threads were extremely helpful! Let me know if there's anything else I should watch out for in this PR
+ https://gist.github.com/shoyer/5f72853c2788e99e785f4737ee8a6ae1 -- JAX wrapper around TF2

\cc @bodono

### CODE REVIEW

**Comment by akshayka** [comment]
This looks great!! Thanks so much for taking this on. I'd love to have this merged in, once it's finished.

**Comment by sbarratt** [comment]
> This looks great!! Thanks so much for taking this on. I'd love to have this merged in, once it's finished.

+1

**Comment by shoyer** [comment]
This is great!

A few ideas for extending this:
- If you want to add support for `jit`, I would suggest looking into jax.experimental.host_callback: https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html
- You might also write the forward operator as a JAX primitive, which would allow you do define a batching rule to support `vmap`.

**Comment by bamos** [comment]
@akshayka @sbarratt, can you do a review through this now? The main layer and all tests are implemented and passing (with GP support), and I'll add the last piece with a sample/tutorial notebook tomorrow.

File: README.md
Comment (akshayka) [inline README.md:5]: At some point we should update the CVXPY Layers logo to include JAX :)

File: cvxpylayers/jax/cvxpylayer.py
Comment (akshayka) [inline cvxpylayers/jax/cvxpylayer.py:80]: At some point we should abstract this into a common function, used by all three layers. Maintaining three separate versions of the same code won't be fun.

No pressure to do this now (though you're more than welcome to do so, if you want to).

**Review by akshayka — APPROVED** [review]
LGTM! It's awesome that we'll finally have a JAX layer.

With this change we now have a lot of duplicated code, across the three layers. It would be great if this code could be unified and re-used by the three layers. You don't need to do that for this PR, though you're more than welcome to do so if you have the inclination.

**Review by sbarratt — APPROVED** [review]
Looks good to me!

File: README.md
Comment (bamos) [reply README.md:5]: Done :)

**Comment by bamos** [comment]
I just pushed a notebook with an intro to the JAX version, squashed the commits, and added followup development issues. I may put in a few last tweaks and will merge in soon once CI passes! Let me know if you all have any other comments or anything else you want to get in before this is merged.

**Comment by froystig** [comment]
To add to @shoyer's comment:
> * If you want to add support for `jit`, I would suggest looking into jax.experimental.host_callback: https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html

Another approach is to use XLA's custom call via jax. [This tutorial](https://github.com/dfm/extending-jax) covers the topic well.

**Comment by akshayka** [comment]
> I just pushed a notebook with an intro to the JAX version, squashed the commits, and added followup development issues. I may put in a few last tweaks and will merge in soon once CI passes! Let me know if you all have any other comments or anything else you want to get in before this is merged.

Awesome! I'll take a look tonight or tomorrow morning. Excited for this to be merged in.

**Pull request merged**
